Refactor virtual connection as a thin wrapper around net.Pipe

bemasc opened the pull request:
The current virtual connection struct `proxiedConn` implements the client side of CONNECT-UDP.  It also contains a great deal of logic to pair reads and writes using a channel, implement deadlines using timers, etc.  Most of this functionality is already available in `masque.Proxy` or `net.Pipe`.

This change reorganizes conn.go and proxy.go to share their implementation of the CONNECT-UDP protocol (which is logically symmetrical for clients and servers), using `net.Pipe` to reverse the direction on the client side.  This reduces the amount of code required in this package and ensures symmetry between the client and server implementations.

The only significant behavior change here is that calls to `ReceiveDatagram(ctx)` no longer pass a nontrivial context, so they cannot time out.  This implies one additional packet of buffering after a read deadline is exceeded.

The main motivation for this change is to enable a sensible implementation of the Capsule Protocol (see draft at https://github.com/bemasc/masque-go/pull/1)

marten-seemann reviewed (commented):
We definitely should be handling capsules correctly!

This PR contains a lot of code, and from a quick glance, lots of changes are only needed for CONNECT-TCP. Can you slim this down to the changes that are required for Capsules as required by RFC 9298?

bemasc committed:
Reduce log spam on connection shutdown

This change adds a function that identifies
errors that are known to be ignorable.  It
applies this check across the proxying
functions to reduce log spam when the proxy
is shut down, or an individual connection is
closed.

This avoids the need to check `s.closed`,
so the proxying logic can be moved from a
method to a function (see next PR).

bemasc committed:
Move proxying logic into direction-independent functions

This allows the proxying logic to be shared between the
client and proxy (see next PR).  This sharing is
important as the proxying logic becomes more complex
with the addition of Capsule support.

bemasc committed:
Refactor virtual connection as a thin wrapper around net.Pipe

Summary:
The current virtual connection struct `proxiedConn` implements the client side of CONNECT-UDP.  It also contains a great deal of logic to pair reads and writes using a channel, implement deadlines using timers, etc.  Most of this functionality is already available in `masque.Proxy` or `net.Pipe`.

This change reorganizes conn.go and proxy.go to share their implementation of the CONNECT-UDP protocol (which is logically symmetrical for clients and servers), using `net.Pipe` to reverse the direction on the client side.  This reduces the amount of code required in this package and ensures symmetry between the client and server implementations.

The only significant behavior change here is that calls to `ReceiveDatagram(ctx)` no longer pass a nontrivial context, so they cannot time out.  This may result in one packet of additional buffering.

The main motivation for this change is to enable a sensible implementation of the Capsule Protocol.
